lib/classes/ConfigSchemaHandler.js

```diff
@@ -0,0 +1,182 @@
+          'Please report the issue at its bug tracker linking:',
+          'https://www.serverless.com/framework/docs/providers/aws/guide/plugins#extending-validation-schema',
+        ].join(' '),
+      ]);
+      this.relaxProviderSchema();
+    }
+
+    const ajv = new Ajv();
+    const validate = ajv.compile(this.schema);
+
+    validate(userConfig);
+    if (validate.errors) {
+      const messages = errorMessageBuilder.buildErrorMessages(validate.errors, userConfig);
+      this.handleErrorMessages(messages);
+    }
+  }
+
+  handleErrorMessages(messages) {
+    if (messages.length) {
+      const errorMessage = messages.join('\n     ');
+      if (this.serverless.service.configValidationMode === 'error') {
+        throw new this.serverless.classes.Error(`${ERROR_PREFIX} ${errorMessage}`);
+      } else {
+        for (const message of messages) {
+          this.serverless.cli.log(`${ERROR_PREFIX} ${message}`, 'Serverless', { color: 'red' });
+        }
+      }
+    }
+  }
+
+  defineTopLevelProperty(name, subSchema) {
+    this.schema.properties[name] = subSchema;
+  }
+
+  defineProvider(name, options = {}) {
+    const currentProvider = this.serverless.service.provider.name;
+    if (currentProvider !== name) {
+      return;
+    }
+
+    this.schema.properties.provider.properties.name = { const: name };
+
+    if (options.provider) {
+      addPropertiesToSchema(this.schema.properties.provider, options.provider);
+    }
+
+    if (options.function) {
+      addPropertiesToSchema(
+        this.schema.properties.functions.patternProperties[FUNCTION_NAME_PATTERN],
+        options.function
+      );
+    }
+
+    if (options.functionEvents) {
+      if (!Array.isArray(options.functionEvents)) throw new Error('functionEvents should be array');
+      for (const functionEvent of options.functionEvents) {
+        this.defineFunctionEvent(name, functionEvent.name, functionEvent.schema);
+      }
+    }
+
+    // In case provider implementers do not set stage of variableSyntax options,
+    // then they are set here. The framework internally sets these options in
+    // Service class, with default values `dev` for stage and
+    // `\\${([^{}]+?)}` for variableSyntax.
+    if (!_.get(this.schema.properties.provider.properties, 'stage')) {
+      addPropertiesToSchema(this.schema.properties.provider, {
+        properties: { stage: { type: 'string' } },
+      });
+    }
+    if (!_.get(this.schema.properties.provider.properties, 'variableSyntax')) {
+      addPropertiesToSchema(this.schema.properties.provider, {
+        properties: { variableSyntax: { type: 'string' } },
+      });
+    }
+  }
+
+  defineCustomProperties(configSchemaParts) {
+    addPropertiesToSchema(this.schema.properties.custom, configSchemaParts);
+  }
+
+  defineFunctionEvent(providerName, name, configSchema) {
+    if (this.serverless.service.provider.name !== providerName) {
+      return;
+    }
+
+    if (
```

Comment:
If we remove it, then when running `sls info` it will crash with error `Cannot read property 'push' of undefined` at ConfigSchmaHandler.defineFunctionEvent. 

The reason I believe is that AWS provider plugin relaxes schema (using `relaxProviderSchema` method), which also includes deleting `__workaround__` event. 

So I believe this condition should exist

Reply:
Sorry, I don't get it. 

Currently, all relaxation happens in one `relaxProviderSchema` which is used two times: one in `awsProvider.js`, and in `configSchemaHandler.validateConfig()`

Also, a concern about the approach with  `__workaround__` came to my mind. I believe there is a conflict of interest. For plugin implementers, it will bring value. But end-users will be confused to see error messages for 'unsupported' events.

Reply:
My point was that we should call `relaxProviderSchema` only in `validateConfig`. It's internal logic doesn't have to change I believe

>  But end-users will be confused to see error messages for 'unsupported' events (they are supported just lack schemas).

It'll only be the case, if provider author will _add_ schema for a provider but would not add of schema for given events, and such looks as an error on provider implementation side.

If provider plugin author wants to take same route as we do with AWS (do not write all schemas upfront), then he should prepare same relaxed schemas as we did here for AWS.

Reply:
When we remove `relaxProviderSchema` from `awsProvider` then we would need to add few lines of code that make `additionalProperties` set to true as such temporary relaxation is still necessary

Reply:
I meant to remove it from `awsProvider` and in `validateConfig` apply it if we're in context of AWS provider.

I believe it wont' require us to add any other logic to `awsProvider` and we also should not change any internals in `relaxProviderSchema`, but condition under which we comment here can be removed.

Reply:
Ah, after a week of break I finally got it. Implemented in https://github.com/serverless/serverless/pull/7335/commits/fe3c0c371b3f47acc4f847caa827f1cc9be4568e